### PR TITLE
Multiple perl packages

### PR DESCRIPTION
**Pull request by inkpot-monkey**
###### Motivation for this change
Adding (and updating) Perl dependencies needed for a project in the summer of NIx.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- 21.11 Release Notes (or backporting 21.05 Release notes)(https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thomasSDK** [comment]
For some reason the github check marks wont let me check "Fits CONTRIBUTING.md."  but instead checks "Tested execution of all binary files (usually in ./result/bin/)"

This should be the other way round

**Comment by stigtsp** [comment]
Hi @thomasSDK, welcome to the NixOS project and thanks for opening the PR.

Can you open a separate PR for `perlPackages.ExtUtilsMakeMaker: 7.56 -> 7.62`? It will make reviewing a bit easier as this module has a few dependents.

**Comment by thomasSDK** [comment]
Thanks @stigtsp good to be here :)

Yep sure no problem will do that now

**Comment by SuperSandro2000** [comment]
> For some reason the github check marks wont let me check "Fits CONTRIBUTING.md." but instead checks "Tested execution of all binary files (usually in ./result/bin/)"
> 
> This should be the other way round

IDK, GitHub is buggy. Just edit the post.

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -11188,6 +11481,34 @@ let
...
+      sha256 = "37835f11ee35326241b4d30368ae1bc195a50414b3662db3e13b865bd52fcde9";
+    };
+    propagatedBuildInputs = [ FileWhich Filechmod LEOCHARREDEBUG Linuxusermod YAML ];
+    meta = {
+      description = "Useful subs for coding cli scripts";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+    };
+  };
+
+  LEOCHARREDEBUG = LEOCHARREDebug;
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:11498]: Isn't this a different CPAN package? https://metacpan.org/dist/LEOCHARRE-DEBUG

(While the above does say its deprecated and prefer to use https://metacpan.org/pod/LEOCHARRE::Debug instead, they're still distinct CPAN dists, which might lead to confusion later on esp when versions are considered.)

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -14028,6 +14421,22 @@ let
     };
   };
 
+  MongoDB = buildPerlPackage {
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:14424]: Maybe worth skipping this?  If upstream itself won't support it anymore, I don't think it should be here.

File: pkgs/top-level/perl-packages.nix [hunk 3]
```diff
@@ -1432,6 +1502,38 @@ let
     };
   };
 
+  BSON = buildPerlPackage {
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:1505]: Ditto for MongoDB upstream stopping support.

File: pkgs/top-level/perl-packages.nix [hunk 4]
```diff
@@ -1432,6 +1502,38 @@ let
...
+    buildInputs = [ JSONMaybeXS PathTiny TestDeep TestFatal ];
+    propagatedBuildInputs = [ CryptURandom Moo TieIxHash boolean namespaceclean ];
+    meta = {
+      homepage = "https://github.com/mongodb-labs/mongo-perl-bson";
+      description = "BSON serialization and deserialization (EOL)";
+      license = lib.licenses.asl20;
+    };
+  };
+
+  BSONXS = buildPerlPackage {
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:1521]: Ditto

**Review by zakame — COMMENTED** [review]
Can I check for the context for these additions, e.g. what project in the Summer of Nix are these for?

**Comment by thomasSDK** [comment]
These additions are to support [openfoodfacts](https://github.com/ngi-nix/openfoodfacts-server) which makes heavy use of mod_perl.

https://github.com/ngi-nix/openfoodfacts-server/blob/main/cpanfile

Wasnt sure what the correct approach was for the modules that upstream stopped supporting, do you reckon that it would be best to keep them local to this SON project? Is there a criteria for adding Perl packages that I should keep in mind?

File: pkgs/top-level/perl-packages.nix (hunk 1 above)
Comment (thomasSDK) [reply pkgs/top-level/perl-packages.nix:11498]: Your right :) for some reason in my head the newer LEOCHARRE::Debug was backwards compatible with the older version but right now I cant find any reason to support my idea.

I'll remove this an for the pull request

**Comment by stigtsp** [comment]
> Wasnt sure what the correct approach was for the modules that upstream stopped supporting [..] 

It's OK in my opinion to include the MongoDB modules (even if they are unfortunately EOL'ed) similar to [other distributions](https://repology.org/project/perl:mongodb/versions). 

> Is there a criteria for adding Perl packages that I should keep in mind?

The convention is that perl packages are the latest version, and is sourced from CPAN - but there are exceptions of course.

File: pkgs/top-level/perl-packages.nix [hunk 5]
```diff
@@ -10445,6 +10445,21 @@ let
...
 
+  ImageOCRTesseract = buildPerlPackage {
+    pname = "Image-OCR-Tesseract";
+    version = "1.26";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/L/LE/LEOCHARRE/Image-OCR-Tesseract-1.26.tar.gz";
+      sha256 = "98d904266a7062f09c9b46f77c4e94529e1fe99339e3f83fda1f92013f007cea";
+    };
+    nativeBuildInputs = [ pkgs.which ];
+    propagatedBuildInputs = [ FileFindRule FileWhich LEOCHARRECLI StringShellQuote pkgs.tesseract pkgs.imagemagick ];
```
[the remark this answers is not in the document]
Comment (zakame) [reply pkgs/top-level/perl-packages.nix]: No worries, glad to help out (and happy you found your solution!) :rocket:

**Comment by zakame** [comment]
Result of `nixpkgs-review pr 133753` [1](https://github.com/Mic92/nixpkgs-review) on x86_64-linux
<details>
  <summary>85 packages built:</summary>
  <ul>
    <li>gnucash</li>
    <li>perl532Packages.ActionCircuitBreaker</li>
    <li>perl532Packages.ActionRetry</li>
    <li>perl532Packages.AlgorithmCheckDigits</li>
    <li>perl532Packages.ApacheDB</li>
    <li>perl532Packages.BSON</li>
    <li>perl532Packages.BSONXS</li>
    <li>perl532Packages.CLDRNumber</li>
    <li>perl532Packages.ColorLibrary</li>
    <li>perl532Packages.DataDumperAutoEncode</li>
    <li>perl532Packages.DevelSize</li>
    <li>perl532Packages.EncodePunycode</li>
    <li>perl532Packages.ExcelWriterXLSX</li>
    <li>perl532Packages.Filechmod</li>
    <li>perl532Packages.FilechmodRecursive</li>
    <li>perl532Packages.FinanceQuote</li>
    <li>perl532Packages.GraphicsColor</li>
    <li>perl532Packages.IOInteractiveTiny</li>
    <li>perl532Packages.ImageOCRTesseract</li>
    <li>perl532Packages.JSONCreate</li>
    <li>perl532Packages.JSONParse</li>
    <li>perl532Packages.LEOCHARRECLI</li>
    <li>perl532Packages.LEOCHARREDebug</li>
    <li>perl532Packages.Linuxusermod</li>
    <li>perl532Packages.LocaleMaketextLexiconGetcontext</li>
    <li>perl532Packages.LogAnyAdapterTAP</li>
    <li>perl532Packages.MathFibonacci</li>
    <li>perl532Packages.ModuleBuildPluggableCPANfile</li>
    <li>perl532Packages.MongoDB</li>
    <li>perl532Packages.MooseXStorage</li>
    <li>perl532Packages.MooseXStorageFormatJSONpm</li>
    <li>perl532Packages.SpreadsheetCSV</li>
    <li>perl532Packages.TestDeepJSON</li>
    <li>perl532Packages.TestDeepType</li>
    <li>perl532Packages.TextFuzzy</li>
    <li>perl532Packages.TimeMoment</li>
    <li>perl532Packages.TypeTinyXS</li>
    <li>perl532Packages.URIEscapeXS</li>
    <li>perl532Packages.UUIDURandom</li>
    <li>perl532Packages.XMLEncoding</li>
    <li>perl532Packages.XMLRules</li>
    <li>perl532Packages.YAMLOld</li>
    <li>perl532Packages.experimental</li>
    <li>perl534Packages.ActionCircuitBreaker</li>
    <li>perl534Packages.ActionRetry</li>
    <li>perl534Packages.AlgorithmCheckDigits</li>
    <li>perl534Packages.ApacheDB</li>
    <li>perl534Packages.BSON</li>
    <li>perl534Packages.BSONXS</li>
    <li>perl534Packages.CLDRNumber</li>
    <li>perl534Packages.ColorLibrary</li>
    <li>perl534Packages.DataDumperAutoEncode</li>
    <li>perl534Packages.DevelSize</li>
    <li>perl534Packages.EncodePunycode</li>
    <li>perl534Packages.ExcelWriterXLSX</li>
    <li>perl534Packages.Filechmod</li>
    <li>perl534Packages.FilechmodRecursive</li>
    <li>perl534Packages.FinanceQuote</li>
    <li>perl534Packages.GraphicsColor</li>
    <li>perl534Packages.IOInteractiveTiny</li>
    <li>perl534Packages.ImageOCRTesseract</li>
    <li>perl534Packages.JSONCreate</li>
    <li>perl534Packages.JSONParse</li>
    <li>perl534Packages.LEOCHARRECLI</li>
    <li>perl534Packages.LEOCHARREDebug</li>
    <li>perl534Packages.Linuxusermod</li>
    <li>perl534Packages.LocaleMaketextLexiconGetcontext</li>
    <li>perl534Packages.LogAnyAdapterTAP</li>
    <li>perl534Packages.MathFibonacci</li>
    <li>perl534Packages.ModuleBuildPluggableCPANfile</li>
    <li>perl534Packages.MongoDB</li>
    <li>perl534Packages.MooseXStorage</li>
    <li>perl534Packages.MooseXStorageFormatJSONpm</li>
    <li>perl534Packages.SpreadsheetCSV</li>
    <li>perl534Packages.TestDeepJSON</li>
    <li>perl534Packages.TestDeepType</li>
    <li>perl534Packages.TextFuzzy</li>
    <li>perl534Packages.TimeMoment</li>
    <li>perl534Packages.TypeTinyXS</li>
    <li>perl534Packages.URIEscapeXS</li>
    <li>perl534Packages.UUIDURandom</li>
    <li>perl534Packages.XMLEncoding</li>
    <li>perl534Packages.XMLRules</li>
    <li>perl534Packages.YAMLOld</li>
    <li>perl534Packages.experimental</li>
  </ul>
</details>

**Review by stigtsp — DISMISSED** [review]
Diff LGTM

**Comment by stigtsp** [comment]
@GrahamcOfBorg eval

**Review by stigtsp — APPROVED** [review]
LGTM. Agree @zakame ?

**Review by zakame — APPROVED** [review]

**Pull request merged**